SourceCode/h8/light.cpp

```diff
@@ -0,0 +1,796 @@
+    }
+    float getY() const {
+        return getField(0, 1);
+    }
+    float getZ() const {
+        return getField(0, 2);
+    }
+    float get(const unsigned int index) const {
+        return getField(0, index);
+    }
+
+    void setX(const float x) {
+        setField(0, 0, x);
+    }
+    void setY(const float y) {
+        setField(0, 1, y);
+    }
+    void setZ(const float z) {
+        setField(0, 2, z);
+    }
+    void set(const unsigned int index, const float value) {
+        setField(0, index, value);
+    }
+
+    void print() const {
+        Matrix::print();
+    }
+
+    Vector &operator=(const Matrix& m) {
+        for(unsigned int i = 0; i < 3; i++) {
+            setField(0, i, m.getField(0, i));
+        }
+        return *this;
+    }
+    Vector &operator=(const float n) {
+        for(unsigned int i = 0; i < 3; i++) {
+            set(i, n);
+        }
+        return *this;
+    }
+
+    Vector operator*(const Vector& v) {
+        Vector result(0, 0, 0);
+        for(unsigned int i = 0; i < 3; i++) {
+            result.set(i, get(i) * v.get(i));
+        }
+        return result;
+    }
+};
+
+namespace Math {
+    Vector add(const Vector a, const Vector b) {
+        return Vector(a.getX() + b.getX(), 
+                        a.getY() + b.getY(), 
+                        a.getZ() + b.getZ());
+    }
+
+    Vector subtract(const Vector a, const Vector b) {
+        return Vector(a.getX() - b.getX(), 
+                        a.getY() - b.getY(), 
+                        a.getZ() - b.getZ());
+    }
+
+    Vector crossProduct(const Vector a, const Vector b) {
+        float x = a.getY() * b.getZ() - a.getZ() * b.getY();
+        float y = a.getZ() * b.getX() - a.getX() * b.getZ();
+        float z = a.getX() * b.getY() - a.getY() * b.getX();
+        return Vector(x, y, z);
+    }
+
+    float dotProduct(const Vector a, const Vector b) {
+        return a.getX() * b.getX() 
+                + a.getY() * b.getY() 
+                + a.getZ() * b.getZ();
+    }
+
+    Matrix matrixMultiply(Matrix a, Matrix b) {
+        Matrix result(a.getRows(), b.getColumns());
+        auto getValue = [&a, &b](const unsigned int i, const unsigned int j) -> float {
+            float result = 0;
+            for(unsigned int add = 0; add < a.getColumns(); add++) {
+                result += a.getField(i, add) * b.getField(add, j);
+            }
+            return result;
+        };
+        for(unsigned int i = 0; i < a.getRows(); i++) {
+            for(unsigned int j = 0; j < b.getColumns(); j++) {
+                result.setField(
+                    i, 
+                    j, 
+                    getValue(i, j)
+                );
+            }
+        }
+        return result;
+    }
+};
+
+#define NUM_OF_TRIANGLE_VERTICES 3
+
+class Triangle {
+    Vector vertex[NUM_OF_TRIANGLE_VERTICES];
+    Vector normal {0, 0, 0};
+public:
+    Triangle(Vector vertex0, Vector vertex1, Vector vertex2) 
```

Comment:
Use references when you pass a Vector, just in case

Reply:
Ok